File: engine/julia/structured_collar_box_arbitrage_engine.jl
# Module BH5 (Julia): Structured Options, Collars & Binary Box Arbitrage Engine
# Quantitative costless collars, Long Box risk-free arbitrage, and binary fixed return payouts in Julia.

module StructuredCollarBoxArbitrageEngine

export structure_collar, evaluate_box, evaluate_binary

function structure_collar(basis::Float64, call_k::Float64, call_prem::Float64, put_k::Float64, put_prem::Float64)
    net_collar = call_prem - put_prem
    is_costless = net_collar >= 0.0
    max_up = (call_k - basis) + net_collar
    max_down = (basis - put_k) - net_collar
    return (net_collar = net_collar, is_costless = is_costless, max_up = max_up, max_down = max_down)
end

function evaluate_box(k1::Float64, k2::Float64, debit::Float64)
    payoff = k2 - k1
    profit = payoff - debit
    return (risk_free_profit = profit, is_profitable = profit > 0.0)
end

function evaluate_binary(bet::Float64, payout_pct::Float64, rebate_pct::Float64, itm::Bool)
    total_return = itm ? bet * (1.0 + payout_pct / 100.0) : bet * (rebate_pct / 100.0)
    net_profit = total_return - bet
    return (total_return = total_return, net_profit = net_profit)
end

end

File: engine/rust/src/structured_collar_box_arbitrage_engine.rs
//! Module BH4 (Rust): Structured Options, Collars & Binary Box Arbitrage Engine.
//! SIMD-optimized costless collars, Long Box risk-free arbitrage, and binary fixed return payouts.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct StructuredCollarBoxState {
    pub net_collar_premium: f64,
    pub max_upside_profit: f64,
    pub max_downside_risk: f64,
    pub box_risk_free_profit: f64,
    pub binary_net_payout: f64,
    pub is_costless_collar: u32,
    pub is_box_profitable: u32,
    pub is_binary_itm: u32,
    pub _padding: [u8; 16],
}

impl StructuredCollarBoxState {
    pub fn evaluate(
        basis: f64, call_k: f64, call_prem: f64, put_k: f64, put_prem: f64,
        box_k1: f64, box_k2: f64, box_debit: f64,
        bet: f64, payout_pct: f64, itm: u32
    ) -> Self {
        let net_collar = call_prem - put_prem;
        let is_costless = if net_collar >= 0.0 { 1 } else { 0 };
        let max_up = (call_k - basis) + net_collar;
        let max_down = (basis - put_k) - net_collar;

        let box_profit = (box_k2 - box_k1) - box_debit;
        let box_profitable = if box_profit > 0.0 { 1 } else { 0 };

        let bin_payout = if itm == 1 { bet * (payout_pct / 100.0) } else { -bet * 0.90 };

        Self {
            net_collar_premium: net_collar,
            max_upside_profit: max_up,
            max_downside_risk: max_down,
            box_risk_free_profit: box_profit,
            binary_net_payout: bin_payout,
            is_costless_collar: is_costless,
            is_box_profitable: box_profitable,
            is_binary_itm: itm,
            _padding: [0u8; 16],
        }
    }
}
